feat: display pokemon details

ChristianAsilo opened the pull request:
- Added pokemon details connector
- Added pokemon details VmFactory and Vm
- Added pokemon details page to display pokemons and its details
- Modefied pokemon action that dispose pokemon details
- Added a new widget for pokemon display view
- Added basic stat card for stat details
- Added new constants in utils
![analyze](https://user-images.githubusercontent.com/128013689/228240103-1287650c-f752-4c98-a4eb-327485553cef.PNG)
![format](https://user-images.githubusercontent.com/128013689/228240114-b23cb795-5730-4734-a208-ac1db635c3e7.PNG)

https://user-images.githubusercontent.com/128013689/228240164-f0da12e1-9916-4430-a21d-1b65e645ccba.mp4

chrstn06 committed:
feat: display pokemon details
- Added pokemon details connector
- Added pokemon details VmFactory and Vm
- Added pokemon details page to display pokemons and its details
- Modefied pokemon action that dispose pokemon details
- Added a new widget for pokemon display view
- Added basic stat card for stat details
- Added new constants in utils

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,25 @@
+    return StoreConnector<AppState, PokemonDetailsVm>(
+      vm: () => PokemonDetailsVmFactory(),
+      onInit: (store) => store.dispatchAsync(GetPokemonDetailsAction(pokemonName: selectedPokemon)),
+      onDispose: (store) => store.dispatchAsync(DisposePokemonDetailAction()),
```
why use `dispatchAsync` here btw? 🤔

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,25 @@
+      onDispose: (store) => store.dispatchAsync(DisposePokemonDetailAction()),
+      builder: (context, vm) => PokemonDetailsPage(
+        pokemonDetails: vm.pokemonDetails,
+        selectedPokemon: selectedPokemon,
```
kindly move this above `pokemonDetails` 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,40 @@
+  }) : super(key: key);
+
+  final Async<PokemonDetails> pokemonDetails;
+  final String selectedPokemon;
```
kindly move this above `pokemonDetails` 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,40 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text(pokemonDetailsText),
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,40 @@
+
+void _showErrorMessageSnackBar(BuildContext context, String? errorMessage) {
+  final SnackBar snackBar = SnackBar(
+    content: Text(errorMessage ?? emptyString),
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_vm.dart:
```diff
@@ -0,0 +1,26 @@
+  Async<PokemonDetails> _pokemonDetails() {
+    if (state.wait.isWaitingFor(GetPokemonDetailsAction.key)) return const Async.loading();
+    if (state.pokemonDetails == null) return const Async.error(errorMessage);
+    return Async(state.pokemonDetails!);
```
kindly add a new line above this 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -22,19 +23,27 @@ class PokemonOverviewPage extends StatelessWidget {
             final pokemon = data[index];
-            return PokemonCard(pokemon: pokemon);
+            return GestureDetector(
+              onTap: () => Navigator.push(
```
kindly move this into a separate private function to make it more readable 🙂

shawn-ffufm commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -22,19 +23,27 @@ class PokemonOverviewPage extends StatelessWidget {
+              onTap: () => Navigator.push(
+                context,
+                MaterialPageRoute(
+                  builder: (context) => PokemonDetailsConnector(selectedPokemon: pokemon.name),
```
kindly rename context to _ since it's unused 🙂

shawn-ffufm commented on lib/state/action/pokemon_actions.dart:
```diff
@@ -33,3 +35,8 @@ class GetPokemonDetailsAction extends ReduxAction<AppState> {
   }
 }
+
+class DisposePokemonDetailAction extends ReduxAction<AppState> {
```
seems u forgot the documentation here? 🤔

shawn-ffufm commented on lib/utils/constants.dart:
```diff
@@ -11,4 +13,23 @@ const appbarTitle = 'Pokedex';
+const abilitiesText = 'Abilities';
+const movesText = 'Moves';
+
+const double defaultTwenty = 20;
```
no need for these up to `defaultSixTeen` 🙂

shawn-ffufm commented on lib/utils/spacing.dart:
```diff
@@ -0,0 +1,25 @@
+    required this.space,
+  }) : super(key: key);
+
+  final double space;
```
kindly rename this to `height` 🙂

shawn-ffufm commented on lib/utils/spacing.dart:
```diff
@@ -0,0 +1,25 @@
+    required this.space,
+  }) : super(key: key);
+
+  final double space;
```
kindly rename this to `width` 🙂

shawn-ffufm commented on lib/widget/basic_stat_card.dart:
```diff
@@ -0,0 +1,62 @@
+              children: [
+                const Text(weightText, style: TextStyle(fontSize: 16)),
+                Text(
+                  '${weight / 10} $kgText',
```
what is the 10 is for? kindly create a constant for it to make it more readable 🙂

shawn-ffufm commented on lib/widget/basic_stat_card.dart:
```diff
@@ -0,0 +1,62 @@
+              children: [
+                const Text(heightText, style: TextStyle(fontSize: 16)),
+                Text(
+                  '${height / 10} $metersText',
```
same here

shawn-ffufm commented on lib/widget/basic_stat_card.dart:
```diff
@@ -0,0 +1,62 @@
+                ),
+              ],
+            ),
+            const VerticalSpace(space: 10),
```
it's supposed to be double not integer value right? kindly apply same to others 🙂

shawn-ffufm commented on lib/widget/pokemon_details_view.dart:
```diff
@@ -0,0 +1,110 @@
+    Key? key,
+  }) : super(key: key);
+
+  final PokemonDetails data;
```
kindly improve the naming here instead of `data` 🙂

shawn-ffufm commented on lib/widget/pokemon_details_view.dart:
```diff
@@ -0,0 +1,110 @@
+              Wrap(
+                spacing: defaultTen,
+                children: data.abilities
+                    .map((ability) => Chip(
```
just use for loop here since it's just a basic ui construction 😝

shawn-ffufm commented on lib/widget/pokemon_details_view.dart:
```diff
@@ -0,0 +1,110 @@
+              SizedBox(
+                child: Wrap(
+                  spacing: defaultTen,
+                  children: data.moves
```
same here 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

chrstn06 committed:
feat: display pokemon details
(pr feedback)
- Modified onDispose to dispatchSync
- Removed extra trailing comma
- Moved GestureDetector to PokemonCard
- Added documentation
- Removed constants
- Improved naming of pokemonDetails instead of data
- Used for loop instead of .map()
- Added pokemon Name and Image in pokemon Details
